### PR TITLE
add vm snapshot for clodutower terraform provider

### PR DESCRIPTION
**Pull request by Sczlog**
- add vm snapshot resource and datasource
- add rollback vm function
- add rebuild vm from snapshot
- update vm resource:
    - set some config as optional, they may get from vm snapshot
    - set optional config as computed as they may get from snapshot (both for vm updation)
    - fix zero value problem
    
TODO: 
- test vm update after default value of cpu_cores and sockets been fixed

example hcl (github not support tf file, use txt to upload it, rename it):
[hcl.txt](https://github.com/smartxworks/cloudtower-terraform-provider/files/8517845/hcl.txt)

### CODE REVIEW

**Comment by Sczlog** [comment]
@netweng @Yuyz0112 PTAL

**Comment by netweng** [comment]
can we add hcl.tf in `examples/` ?

File: internal/provider/data_source_vm_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	if name := d.Get("name").(string); name != "" {
+		gp.RequestBody.Where.Name = &name
+	} else if nameContains := d.Get("name_contains").(string); nameContains != "" {
+		gp.RequestBody.Where.NameContains = &nameContains
+	}
+	if vm_id := d.Get("vm_id").(string); vm_id != "" {
+		gp.RequestBody.Where.VM.ID = &vm_id
+	}
+	clusters, err := ct.Api.VMSnapshot.GetVMSnapshots(gp)
```
Comment (netweng) [inline internal/provider/data_source_vm_snapshot.go:84]: clusters ?

File: main.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ func main() {
 	opts := &plugin.ServeOpts{ProviderFunc: provider.New(version)}
 
 	if debugMode {
-		err := plugin.Debug(context.Background(), "registry.terraform.io/smartx/cloudtower", opts)
+		err := plugin.Debug(context.Background(), "registry.terraform.io/yuyz0112/cloudtower", opts)
```
Comment (netweng) [inline main.go:40]: should be smartx ?

File: internal/provider/data_source_vm_snapshot.go (hunk 1 above)
Comment (Sczlog) [reply internal/provider/data_source_vm_snapshot.go:84]: updated

File: main.go (hunk 2 above)
Comment (Sczlog) [reply main.go]: It need change makefile to fix the path.

**Comment by Sczlog** [comment]
@netweng hcl is updated to snapshot.tf in the example folder

**Comment by netweng** [comment]
LGTM

**Pull request merged**